desc/objectives/_coils.py

```diff
@@ -0,0 +1,472 @@
+import numpy as np
+
+from desc.backend import jnp
+from desc.compute import compute as compute_fun
+from desc.compute import get_profiles, get_transforms, rpz2xyz, rpz2xyz_vec
+from desc.compute.utils import safenorm
+from desc.grid import LinearGrid
+from desc.utils import Timer, warnif
+
+from .objective_funs import _Objective
+
+
+class QuadraticFlux(_Objective):
+    """Target the quadratic flux on an equilibrium from a magnetic field.
+
+    compute
+
+    (B.n)^2
+
+    where n is the normal vector to the plasma surface, and B is the magnetic field at
+    the plasma surface.
+
+    NOTE: Only works for vacuum equilibria currently
+
+    Parameters
+    ----------
+    eq : Equilibrium, optional
+        Equilibrium that will be optimized to satisfy the Objective.
+    field : MagneticField
+        MagneticField object, the parameters of this will be optimized
+        to minimize the objective.
+    target : {float, ndarray}, optional
+        Target value(s) of the objective. Only used if bounds is None.
+        Must be broadcastable to Objective.dim_f.
+    bounds : tuple of {float, ndarray}, optional
+        Lower and upper bounds on the objective. Overrides target.
+        Both bounds must be broadcastable to to Objective.dim_f
+    weight : {float, ndarray}, optional
+        Weighting to apply to the Objective, relative to other Objectives.
+        Must be broadcastable to to Objective.dim_f
+    normalize : bool, optional
+        Whether to compute the error in physical units or non-dimensionalize.
+        Note: has no effect on this objective.
+        FIXME: add normalization for the B part of this objective
+    normalize_target : bool
+        Whether target should be normalized before comparing to computed values.
+        if `normalize` is `True` and the target is in physical units, this should also
+        be set to True.
+        Note: has no effect on this objective.
+    loss_function : {None, 'mean', 'min', 'max'}, optional
+        Loss function to apply to the objective values once computed. This function
+        is called on the raw compute value, before any shifting, scaling, or
+        normalization. Note: has no effect for this objective
+    deriv_mode : {"auto", "fwd", "rev"}
+        Specify how to compute jacobian matrix, either forward mode or reverse mode AD.
+        "auto" selects forward or reverse mode based on the size of the input and output
+        of the objective. Has no effect on self.grad or self.hess which always use
+        reverse mode and forward over reverse mode respectively.
+    source_grid : Grid, optional
+        Collocation grid containing the nodes to evaluate field source at on
+        the winding surface. (used if e.g. field is a CoilSet or
+        FourierCurrentPotentialField)
+    eval_grid : Grid, optional
+        Collocation grid containing the nodes to evaluate the normal magnetic field at
+        plasma geometry at.
+    external_field : MagneticField, optional
```

Comment:
Why do we need 2 fields?

Reply:
This can be removed once we make sure optimizable collections works (as then we can pass in a SumMagneticField object with the TF field parameters fixed), but this is to allow optimization of say a modular coilset while also taking into consideratin an external TF coilset or a simple 1/R toroidal field